### DISCUSSION TITLE
docs: Add a new variant (board) and "Generic System Clock configuration"

### DISCUSSION THREAD

**Discussion by thernstig** [stm32duino/Arduino_Core_STM32#2563] [WiKi]
I had a hard time understanding https://github.com/stm32duino/Arduino_Core_STM32/wiki/Add-a-new-variant-%28board%29#3---generic-system-clock-configuration.

I am quoting the parts I found hard and what I felt was missing/strange:

> Go to Pinout tab, enable the desired peripherals which require specific clock configuration (not needed for peripherals clocked by HCLKx, or APBx clock): SDIO, USB, ... In this example only USB needs to be enabled as other peripherals default clock are correct by default.

What is "desired peripherals" here? How would I know what to enable?

> Set PLL Source Mux to HSI.

There is no `PLL Source Mux` for the MCU I added in https://github.com/stm32duino/Arduino_Core_STM32/pull/2556. The sentence is written like it is something that should exist, but it does not seem true for all MCUs.

**Comment by thernstig** [comment]
In addition, this part:

> 6 - Restart
> Restart Arduino IDE and try one of the new entry with the [CheckVariant example](https://github.com/stm32duino/STM32Examples/tree/main/examples/NonReg/CheckVariant).

It mentions nothing on how to do this on that above page. Should there have been a link to somewhere on how to load the new board into Arduino to test it?

**Comment by fpistm** [comment]
The board should appears in the menu you add it that's all. Select it and voilà.

**Comment by thernstig** [comment]
@fpistm I am talking about an unreleased version of Arduino_Core_STM32. Then the new board does not immediately show up in the JSON file does it?

**Comment by fpistm** [comment]
Yes. I need to perform a release.
But if you use the git repository in the Arduino package you will see it.

**Comment by thernstig** [comment]
@fpistm I think this part:

> But if you use the git repository in the Arduino package you will see it.

Is not explained from https://github.com/stm32duino/Arduino_Core_STM32/wiki/Add-a-new-variant-%28board%29#6---restart.

How would one know how to do this? I just started with both Arduino and STM32, so consider it from a point from someone who never used these tools before (even though understanding the basics).

**Comment by fpistm** [comment]
It is explained here: https://github.com/stm32duino/Arduino_Core_STM32/wiki/Using-git-repository

Feel free to propose update, it is a community project.

**Comment by thernstig** [comment]
I would propose a lot more if I knew the answers 😆

**Comment by thernstig** [comment]
Wrote https://github.com/stm32duino/Arduino_Core_STM32/issues/2559 as a separate issue.

The two core issues of this topic are still valid. I do not know how to update those as I have no idea what those parts in reality mean.

**Comment by thernstig** [comment]
@fpistm Not sure why this was converted into a discussion. The Wiki is ambiguous and mentions things that do not exist for all MCUs in STM32CubeMx.

**Reply by fpistm** [reply]
As far as I know, it is the first series without PLL mux. At the time it was written it was not the case. That's all.

**Reply by thernstig** [reply]
Got it. This means the wiki should be updated though? 🙂

**Reply by fpistm** [reply]
Hi @thernstig 

I've fully reworked the wiki page to add a generic board:
https://github.com/stm32duino/Arduino_Core_STM32/wiki/Add-a-new-variant-%28board%29
and also this one: 
https://github.com/stm32duino/Arduino_Core_STM32/wiki/Using-git-repository

Let me know if it's ok for you.

**Reply by thernstig** [reply]
@fpistm does this mean https://raw.githubusercontent.com/stm32duino/BoardManagerFiles/dev/package_stmicroelectronics_index.json can be used directly and everything should work according to https://github.com/stm32duino/Arduino_Core_STM32/wiki/Using-git-repository?

It would be great that if someone selects that JSON file, all the proper dependencies, as they are on the main branch, are downloaded automatically by Arduino.

Regarding the generic board page, there are some things I find a bit ambiguous, but I think it's ok. It is hard giving feedback here since there are certain sections/passages that I would update, and I might also lack knowledge to myself properly edit/propose a change without a proper back/forth discussion unfortunately.

**Reply by fpistm** [reply]
@thernstig if you use the package index from the `dev` branch of  **BoardManagerFiles** repo then it will ensure to have the correct tools dependencies but the core package is the same than the one provided with the latest release.
If no tools dependencies are required to use the `main` branch of the **Arduino_Core_STM32** then `dev` and `main` branch of  BoardManagerFiles repo are the same (same json file).
If tools dependencies are required then json file will differ and if you simply install it without using the `main` branch of the **Arduino_Core_STM32** using git but the installed one then you probably can't compile as the dependency will not be satisfied.

**Reply by fpistm** [reply]
> I find a bit ambiguous,

Do not forget that your are in the **"Advanced usages"** section of the wiki. :wink: 
Currently, I'm updating the page and added this:
> [!TIP]
> If you have any issue with the following guide. Do not hesitate to ask question on the [stm32duino Github discussions](https://github.com/orgs/stm32duino/discussions) or the [stm32duino forum](http://stm32duino.com).

**Reply by fpistm** [reply]
@thernstig 
I've reword the usage of the dev json file. Hope it is more clear.

**Reply by thernstig** [reply]
@fpistm understood I am in the advanced section, but I still think things can improve 😄 

It is more clear. But I do not understand why you list all dependencies and have this text:

> To get started with development on main branch, correct versions of the required dependencies have to be installed

If the only thing one has to do to install the latest dependencies in the Arduino IDE is to use the https://raw.githubusercontent.com/stm32duino/BoardManagerFiles/dev/package_stmicroelectronics_index.json file, then I see no meaning to list all those dependencies. The wiki's purpose must be to just show quickly how to use the latest boards on main?

I would also rephrase the title possibly. I think "Using git repository" could be interpreted in many ways. It would be clearer maybe to say "How to use the latest unreleased version of STM32duino". That it is done by cloning the git repo is an implementation detail.

The best thing would of course be if you could setup the Github workflows so the artifact https://raw.githubusercontent.com/stm32duino/BoardManagerFiles/dev/package_stmicroelectronics_index.json was updated to point to the latest parts from this repo so the only thing users had to do would be to add the https://raw.githubusercontent.com/stm32duino/BoardManagerFiles/dev/package_stmicroelectronics_index.json in the "Additional boards manager URLs" in the Arduino IDE, and not have to do anything else 🙂 And even build such a .json file per commit, so people can test their own commits.

**Reply by fpistm** [reply]
I will rephrase as suggested.
About Artifact and json generation (daily or weekly) is something I've already considered, but time is missing for that.
This could be fine for other users to test but the main goal of using the git repo is also to develop/test before to provide a PR and so you have to use the git repo.

**Reply by fpistm** [reply]
And thanks a lot for all your great suggestions and contributions. Very appreciated. 👏

**Reply by thernstig** [reply]
Got it. Well I think you can leave it as-is then now. Better focus in the future on seeing if each commit can build a JSON file. Would be much more testable. Maybe even testable via the Ardunio CLI directly in CI!

**Reply by fpistm** [reply]
Ci already use the dev json file else it will not build when dependencies are updated. :wink:
